app/views/contests/scoreboard.html.erb

```diff
@@ -29,20 +39,25 @@
             (<%= @weighting[problem.id] %>)
           </th>
         <% end %>
-        <th style="text-align: right"> Score </th>
-        <th> Time </th>
-        <th> Rank </th>
+        <th style="width:1%; text-align: right"> Score </th>
+        <th style="width:1%"> Time </th>
       </tr>
     </thead>
     <tbody>
-      <% median = @scoreboard[@scoreboard.length/2-1] %>
       <% rank = 1 %>
+      <% num_unofficial = 0 %>
       <% previous_record = @scoreboard.first %>
       <% @scoreboard.each_with_index do |record,index| %>
-        <% if !user_signed_in? or record.user && record.user.id != current_user.id && !policy(@contest).inspect? %>
-          <% next if record.score < median.score || (record.score == median.score && record.time_taken > median.time_taken) # no permission to view %>
-        <% end %>
-        <tr <% if user_signed_in? && record.user && record.user.id == current_user.id %> class="emphasized"<% end %>>
+        <% next if !show_scoreboard && (!user_signed_in? or record.user && record.user.id != current_user.id) # no permission to view %>
+        <% is_unofficial = @contest.show_unofficial_contestants && (record.school.nil? || record.school_year.nil?) %>
+        <tr class="<% if user_signed_in? && record.user && record.user.id == current_user.id %>emphasized<% end %><% if is_unofficial %> unofficial<% end %>">
+          <% num_unofficial += 1 if is_unofficial %>
+          <% rank = index + 1 - num_unofficial unless previous_record.score == record.score && previous_record.time_taken == record.time_taken %>
```

Comment:
I think there is a bug here. I haven't tested it, but I'm pretty sure the following scoreboard will trigger it:

| Rank | User           | Score |
|------|----------------|------:|
| 1    | A              | 100   |
| -    | B (unofficial) | 90    |
| 2    | C              | 90    |

I'm pretty sure the rank of user C will be (incorrectly) displayed as 1, because the rank isn't updated when the score is equal to the previous score.

How about:

```erb
# before loop
      <% num_ranked = 0 %>
      <% rank = nil %>
      <% previous_record = nil %>
# in loop
          <% if is_ranked %>
            <% num_ranked += 1 %>
            <% rank = num_ranked unless previous_record && previous_record.score == record.score && previous_record.time_taken == record.time_taken %>
            <% previous_record = record %>
          <% end %>
```

Reply:
Good catch.